src/test/java/org/apache/sysml/test/gpu/NeuralNetworkOpTests.java

```diff
@@ -0,0 +1,490 @@
+														if (filterSize > MAX_OP_SIZE)  // filter size
+															continue;
+
+														int P = (int) ConvolutionUtils.getP(H, R, strideH, padH);
+														int Q = (int) ConvolutionUtils.getQ(W, S, strideW, padW);
+
+														long doutSize = N * K * P * Q * 8l;
+														if (doutSize > MAX_OP_SIZE) // dout/output size
+															continue;
+
+														double imageSizeInMB = imageSize / (1024.0 * 1024.0);
+														double filterSizeInMB = filterSize / (1024.0 * 1024.0);
+														double doutSizeInMB = doutSize / (1024.0 * 1024.0);
+														System.out
+																.format("conv2d_backward_data, image[%d,%d,%d,%d](%.1fMB), filter[%d,%d,%d,%d](%.1f), dout[%d,%d,%d,%d](%.1fMB), stride[%d,%d], padding[%d,%d]",
+																		N, C, H, W, imageSizeInMB, N, C, R, S,
+																		filterSizeInMB, N, K, P, Q, doutSizeInMB,
+																		strideH, strideW, padH, padW);
+
+														Matrix filter = generateInputMatrix(spark, (int) K,
+																(int) (C * R * S), sparsity, seed);
+														Matrix dout = generateInputMatrix(spark, (int) N,
+																(int) (K * P * Q), sparsity, seed);
+														HashMap<String, Object> inputs = new HashMap<>();
+														inputs.put("N", N);
+														inputs.put("C", C);
+														inputs.put("H", H);
+														inputs.put("W", W);
+														inputs.put("K", K);
+														inputs.put("R", R);
+														inputs.put("S", S);
+														inputs.put("strideH", strideH);
+														inputs.put("strideW", strideW);
+														inputs.put("padH", padH);
+														inputs.put("padW", padW);
+														inputs.put("filter", filter);
+														inputs.put("dout", dout);
+														List<Object> outCPU = runOnCPU(spark, scriptStr, inputs,
+																Arrays.asList("O"));
+														List<Object> outGPU = runOnGPU(spark, scriptStr, inputs,
+																Arrays.asList("O"));
+														assertHeavyHitterPresent("gpu_conv2d_backward_data");
+														assertEqualObjects(outCPU.get(0), outGPU.get(0));
+														clearGPUMemory();
+													}
+												}
+											}
+										}
+									}
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	@Test
+	public void testMaxPool() {
+		String scriptStr = "O = max_pool(image, padding=[padH, padW], stride=[strideH, strideW], input_shape=[N,C,H,W], pool_size=[R,S])";
+
+		for (long N : Nlst) {
+			for (long C : Clst) {
+				for (long H : Hlst) {
+					for (long W : Wlst) {
+						for (long R : Rlst) {
+							for (long S : Slst) {
+								for (long strideH : strideHeightLst) {
+									for (long strideW : strideWidthLst) {
+										for (long padH : padHeightLst) {
+											for (long padW : padWidthLst) {
+												for (double sparsity : sparsitylst) {
+
+													// pool is smaller than image + padding
+													if (R > (H + padH) || S > (W + padW))
```

Comment:
Similar to conv2d, the max pool window could be equal in size to the image + pad.